lib/engine/game/g_system18/map_britain_customization.rb

```diff
@@ -0,0 +1,497 @@
+        end
+
+        def map_britain_remove_icon(entity, hex_id)
+          return unless entity.corporation?
+          return if @britain_corps_with_mines[entity.name]
+
+          hex = hex_by_id(hex_id)
+          hex.tile.icons.clear # should be the only icon on the tile
+          @log << "#{entity.name} takes mine token from #{hex_id}"
+          @britain_corps_with_mines[entity.name] = hex_id
+        end
+
+        def map_britain_removable_icon_action_str
+          'Take a mine token'
+        end
+
+        def map_britain_extra_revenue_for(route, stops)
+          map_britain_bonuses(route, stops)[:revenue]
+        end
+
+        def map_britain_extra_revenue_str(route)
+          bonus = map_britain_bonuses(route, route.stops)[:description]
+          bonus ? " + #{bonus}" : ''
+        end
+
+        def map_britain_bonuses(route, stops)
+          train = route.train
+          bonus = { revenue: 0 }
+          return bonus unless train
+
+          desc = ''
+
+          # mines (no 4D doubling), only on one run
+          corp = train.owner
+          lowest_train = route.routes.reject { |r| r.stops.empty? }.min_by { |r| corp.trains.index(r.train) }.train
+          if (train == lowest_train) && corp && @britain_corps_with_mines[corp.name]
+            bonus[:revenue] += BRITAIN_MINE_BONUS[@phase.name.to_i - 2]
+            desc = 'Mine'
+          end
+
+          # Scotland (doubles with 4D)
+          scotland_city = stops.find do |stop|
+            BRITAIN_REGION_HEXES['Scotland'].include?(stop.hex.id) && (stop.city? || stop.offboard?)
+          end
+          non_scotland_city = stops.find { |stop| !BRITAIN_REGION_HEXES['Scotland'].include?(stop.hex.id) && stop.city? }
+          if scotland_city && non_scotland_city
+            bonus[:revenue] += BRITAIN_SCOTLAND_BONUS[@phase.name.to_i - 2] * (train.name == '4D' ? 2 : 1)
+            desc += ', ' unless desc == ''
+            desc += 'Scotland'
+          end
+
+          # Wales (doubles with 4D)
+          wales_city = stops.find { |stop| BRITAIN_REGION_HEXES['Wales'].include?(stop.hex.id) && (stop.city? || stop.offboard?) }
+          non_wales_city = stops.find { |stop| !BRITAIN_REGION_HEXES['Wales'].include?(stop.hex.id) && stop.city? }
+          if wales_city && non_wales_city
+            bonus[:revenue] += BRITAIN_WALES_BONUS[@phase.name.to_i - 2] * (train.name == '4D' ? 2 : 1)
+            desc += ', ' unless desc == ''
+            desc += 'Wales'
+          end
+
+          # London-Port (doubles with 4D)
+          london = stops.find { |stop| stop.hex.id == BRITAIN_LONDON_HEX }
+          port = stops.find { |stop| stop.tile.icons.any? { |i| i.name == 'port' } }
+          if london && port
+            bonus[:revenue] += BRITAIN_LONDON_BONUS[@phase.name.to_i - 2] * (train.name == '4D' ? 2 : 1)
+            desc += ', ' unless desc == ''
+            desc += 'London-Port'
+          end
+
+          bonus[:description] = "(#{desc})" unless desc == ''
```

Comment:
Rather than conditionally adding these commas and spaces, you could set `desc` to be an empty array at the start of the method, and then add a new string for each bonus, then join them later.
```
          desc = []
          …

          if london && port
            bonus[:revenue] += BRITAIN_LONDON_BONUS[@phase.name.to_i - 2] * (train.name == '4D' ? 2 : 1)
            desc << 'London-Port'
          end

          bonus[:description] = "(#{desc.join(', ')})" unless desc.empty?
```

Reply:
OK, I like that.